README.adoc
:toc:
:toclevels: 4

= Wayk Den PowerShell Cmdlet

WaykDen-ps is a PowerShell cmdlet for Wayk Den, the centralized server for https://wayk.devolutions.net[Devolutions Wayk Now]. For assistance or feature requests, please use the https://forum.devolutions.net/#WaykNow[Wayk Now forums] or open a ticket on this github project.

Make sure you have all software prerequisites (<<Installation>> section) then read one of our getting started guides:

* link:docs\getting-started-argo.adoc[getting started guide (Argo tunnel)]
* link:docs\getting-started-azure.adoc[getting started guide (Azure VM)].
* link:docs\getting-started-relay.adoc[getting started guide (Relay servers)].
* link:docs\getting-started-acme.adoc[getting started guide (ACME/letsencrypt)].

Once you have been familiarized with Wayk Den, come back to this page for more details on specific aspects of the software.

== Installation

=== PowerShell Core

On Windows, the regular PowerShell already installed can be used. On macOS and Linux, you need PowerShell Core 6.0:

PowerShell Core 6.0::
https://github.com/PowerShell/PowerShell#get-powershell

To launch a PowerShell Core terminal, use the `pwsh` command.

=== Wayk Den PowerShell module

A powershell module is available to configure and manage the Wayk Den server. It is available on PSGallery or it could be built directly from https://github.com/Devolutions/WaykDen-ps[source code]. We recommand to install the latest version available on PSGallery, but if needed, a section describe how to build it if you need it.

==== Installation from PSGallery

The Wayk Den PowerShell module is https://www.powershellgallery.com/packages/WaykDen[available on PSGallery].

[source,sh]
----
Install-Module -Name WaykDen
Import-Module WaykDen
----

If you encounter issues with the `Install-Module` command, you may have to https://docs.microsoft.com/en-ca/powershell/gallery/installing-psget[install or update PowerShellGet].

You can then list all commands exported from the `WaykDen` module:

[source,sh]
----
Get-Command -Module WaykDen
----

When a new version comes out, you can update the cmdlet and restart Wayk Den to upgrade. The updated cmdlet will point to newer versions of the docker containers used in Wayk Den, so upgrading is a simple matter of restarting Wayk Den:

[source,sh]
----
Update-Module -Name WaykDen
Import-Module WaykDen
Restart-WaykDen
----

==== Installation from Source

If you have installed the module from PSGallery, this section is not useful for you. You can jump to the next section. 

To build the Cmdlet, you will need the .NET Core 2.2 SDK and the .NET Framework 4.7.2 SDK. The latter is only needed if you want to support PowerShell for Windows.

 .NET Core 2.2 SDK::
https://dotnet.microsoft.com/download

 .NET Framework 4.7.2 SDK::
https://dotnet.microsoft.com/download/dotnet-framework/net472

From PowerShell, run the build.ps1 script, and then load the resulting PowerShell module:

[source,sh]
----
.\build.ps1
Import-Module "./WaykDen" -Force
----

=== Docker Installation

Wayk Den is composed of multiple microservices running inside Docker containers. The PowerShell cmdlet simplifies the usage of Docker for users not familiar with it. For more advanced users, the cmdlet can export a usable docker-compose configuration file.

==== Windows Server 2019

The official way of running Wayk Den on Windows Server is to use the built-in support for Windows containers instead of installing Docker Desktop for Windows and running Linux containers. If you intend to run Wayk Den inside a virtual machine, using Windows Server and Windows containers is the recommended approach because it does not require Docker Desktop for Windows and a Hyper-V virtual machine.

Install or update NuGet and PowerShellGet:
----
$ Install-PackageProvider -Name NuGet -Force
$ Install-Module -Name PowerShellGet -Force
----

Check if the Windows Server 'Containers' feature is installed, then install it:
----
$ Get-WindowsFeature | Where-Object { $_.Name -Like 'Containers' }
$ Install-WindowsFeature -Name Containers -Restart
----

Install the Microsoft Docker provider:
----
$ Install-Module -Name DockerMsftProvider -Force
$ Install-Package -Name docker -ProviderName DockerMsftProvider -Force
----

Start the docker service if it is not currently running:
----
$ Get-Service | Where-Object { $_.Name -Like 'docker' }
$ Start-Service -Name docker
----

==== Windows 10

On Windows 10, the recommended option is to use https://hub.docker.com/editions/community/docker-ce-desktop-windows[Docker Desktop for Windows]. This version of Docker requires Hyper-V for Linux container support, making it suitable for testing on a physical machine, but not so much for a virtual machine because of limited support of Docker Desktop for Windows in virtualized environments. If you want to try Wayk Den inside a virtual machine, use Windows Server 2019 instead with Windows containers.

For bind mount support with Linux containers on Windows, you need to go in the Docker settings and https://rominirani.com/docker-on-windows-mounting-host-directories-d96f3f056a2c[select the required drives from the "Shared Drives" section].

==== macOS

On macOS, follow the official https://docs.docker.com/docker-for-mac/[Docker Desktop getting started guide].

==== Linux

On Linux, follow one of the https://docs.docker.com/install/linux/docker-ce/ubuntu/[distribution-specific getting started guides], then do not forget to https://docs.docker.com/install/linux/linux-postinstall/[add your user to the docker group] after installation.

To confirm that Docker is up and running, use the `docker run hello-world` command. If you don't see the "Hello from Docker!" message, something is wrong with your Docker installation.

== Configuration

All Wayk Den commands accept a -ConfigPath parameter to point to a directory where various configuration files are kept. If this parameter is not specified, the current working directory is used by default.

The first step is to create a Wayk Den configuration file with mandatory parameters using the `New-WaykDenConfig` command.

You need to choose a realm for your Wayk Den, such as "contoso.net". This realm does not need to be a valid DNS domain name, but it will become your Wayk Den realm used in various places, including the peer-to-peer certificates generated by the Wayk Den built-in certificate authority.

The second mandatory parameter is the external URL at which the Wayk Den will be exposed. We recommend using the "den" subdomain under the domain of your choice, such as "den.contoso.net". The protocol prefix (`http://` or `https://`) also needs to be included.

Create a new Wayk Den configuration file, using "contoso.net" as realm and "https://den.contoso.net" as the external url:

[source, sh]
----
PS > New-WaykDenConfig -Realm contoso.net -ExternalUrl https://den.contoso.net -ListenerUrl http://localhost:4000
----

The listener url is used to specify the listening protocol and port on the machine where Wayk Den is running. To listen in https, a certificate needs to be imported before changing the listener URL protocol.

=== Certificate configuration

==== Using certificates with Wayk Den

To expose Wayk Den to the external world, you will need a TLS certificate from a trusted certificate authority such as https://letsencrypt.org/[Let's Encrypt]. Wayk Now will validate certificates in the same way as a browser does for a website.

If you wish to make your own certificate authority, the Root CA certificate will need to be installed in each machine's certificate trust store. If the system browser can validate it, Wayk Now should be able to validate it as well, it's just a lot more work.

Before going any further, check the following points:

1. The certificate name matches your configured external URL. This means that if your external URL is "https://den.contoso.com" then your certificate name should be "den.contoso.com" or "*.contoso.com" if you have a wildcard certificate.

2. The certificate is either in pem + key or pfx (PKCS#12) format. The private key password is only supported for the pfx format.

3. The certificate file contains the certificate *chain* excluding the Root CA. This means that in most cases, you should have a leaf certificate, followed by one or more intermediate certificates. If validation works in a browser but not in Wayk Now, the intermediate certificate is likely missing.

[[certificate-command]]The PEM format is the simplest to work with, since it is the Base64 representation of the DER-encoded certificate in between "-----BEGIN CERTIFICATE-----" and "-----END CERTIFICATE-----" tags. To add the intermediate certificate to the certificate file, just append it after your leaf certificate in a text editor.

[source, sh]
----
PS > Import-WaykDenCertificate -CertificateFile /path/to/certificate.pem -PrivateKeyFile /path/to/private_key.key
----

[source, sh]
----
PS > Import-WaykDenCertificate -CertificateFile /path/to/certificate.pfx -Password <password>
----

=== Relay Configuration

To get started, it is recommended to try using our existing relay servers, even if you intend to deploy your own servers later on. It is the quickest path to getting something functional up and running.

Set your Jet Relay Url to `https://api.jet-relay.net`:

[source, sh]
----
PS > Set-WaykDenConfig -JetRelayUrl https://api.jet-relay.net
----

Continue your initial configuration, and once you know it is working, come back to the link:docs\getting-started-relay.adoc[Devolutions Jet] page for more information on how to deploy your own relays.

=== System Service

It is possible to register a system service on Windows to automatically call Start-WaykDen / Stop-WaykDen in a given config directory.

Instead of using a local user directory for the configuration files, the service should be installed in a global directory with restricted permissions like this:

[source, sh]
----
PS > $ConfigPath = Get-WaykDenPath ConfigPath
PS > $ConfigPath
C:\ProgramData\Devolutions\Wayk Den
PS > New-Item -Path $ConfigPath -Type Directory -Force
----

The "Get-WaykDenPath ConfigPath" command returns the recommended directory for the current platform where files should be stored.

If you already have a directory with your Wayk Den configuration files, you should move them to the new location. You can then proceed with the service registration:

[source, sh]
----
PS > Set-Location $ConfigPath
PS > Register-WaykDenService -ServicePath $ConfigPath
"WaykDen" service has been installed to "C:\ProgramData\Devolutions\Wayk Den"
PS > Get-Service WaykDen

Status   Name               DisplayName
------   ----               -----------
Stopped  WaykDen            Wayk Den
PS > Start-Service WaykDen
----

Wait for the service to start, this can take some time. If you look at the files in the directory, you should see "WaykDen.exe", which is the service executable, and "WaykDen.log", the service log file:

[source, sh]
----
PS > Get-ChildItem


    Directory: C:\ProgramData\Devolutions\Wayk Den


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----         4/2/2020   5:21 PM                den-server
d-----         4/2/2020   5:06 PM                traefik
-a----         4/7/2020   4:34 PM            314 service.json
-a----         4/2/2020   5:06 PM            909 wayk-den.yml
-a----         4/2/2020   5:33 PM         770048 WaykDen.exe
-a----         4/7/2020   4:07 PM           2753 WaykDen.log
----

If you encounter issues with Wayk Den not working as expected when launched through the system service, the best way to debug the issue is to go in the same directory as the service and run the "Start-WaykDen" command manually to see what happens.

The system service is a wrapper over the cmdlet. If you need to remove it, you can call the 'Unregister-WaykDenService' function:

[source, sh]
----
PS > Unregister-WaykDenService -ServicePath $ConfigPath
----

You can also update the service wrapper executable when never versions of the cmdlet are published by calling Register-WaykDenService again.

== Data Persistence

Aside from the Wayk Den configuration files, all Wayk Den data is stored inside a https://www.mongodb.com[MongoDB database]. If you do not specify a database, a simple MongoDB container will be launched with a docker volume attached to it. While this is fine for a lot of use cases, using a single MongoDB container instance is not suitable for high availability scenarios.

To point Wayk Den to an existing MongoDB server, you need to configure a database connection string, and set MongoDB as external to avoid launching a MongoDB container instance:

----
PS > Set-WaykDenConfig -MongoExternal $true -MongoUrl 'mongodb://mongo-server:27017'
----

=== Database Backup

MongoDB database backups can be done using the https://docs.mongodb.com/manual/reference/program/mongodump/[`mongodump` command-line tool]. For simplicity, the `Backup-WaykDenData` command can be used to streamline the process:

----
PS > Backup-WaykDenData -BackupPath .\den-backup.tgz -Verbose
----

The `Backup-WaykDenData` is only usable with a non-external MongoDB database. You can add the `-Verbose` parameter to show the docker commands used to call `mongodump` inside a container for reference.

=== Database Restore

A MongoDB database can be restored from a backup using the https://docs.mongodb.com/manual/reference/program/mongorestore/[`mongorestore` command-line tool]. For obvious reasons, you should be careful about restoring a database, as the operation will overwrite the current data and replace it with the one from the backup.

Just like `Backup-WaykDenData`, `Restore-WaykDenData` should only be used with a non-external MongoDB database. A proper restore involves stopping Wayk Den, restoring data from the backup file, then restarting Wayk Den. Since Stop-WaykDen stops the database container, `Restore-WaykDenData` will automatically start it for the restore operation.

----
PS > Stop-WaykDen
PS > Restore-WaykDenData -BackupPath .\den-backup.tgz -Verbose
PS > Restart-WaykDen
----

== Running

Start Wayk Den, and wait for all microservices to start:

[source, sh]
----
PS > Start-WaykDen
----

Once started, Wayk Den listens on http://localhost:4000 by default. We recommend using a reverse tunnel such as https://ngrok.com/[ngrok] or https://www.cloudflare.com/en-ca/products/argo-tunnel/[argo tunnels] from Cloudflare. In this case, a tunnel is used to expose localhost:4000 on the den.contoso.net external url.

You can check that all containers are up and running with the `docker ps -f network=den-network` command.

To confirm everything is correctly configured, you should be able to get a response from the Wayk Den well known configuration endpoint:

[source, sh]
----
curl http://localhost:4000/.well-known/configuration
{"den_router_uri":"https://den.contoso.net/cow","lucid_uri":"https://den.contoso.net/lucid","realm":"contoso.net","wayk_client_id":"zqdvSbCRWdDrj1fQXwzPQbCg"}
----

If you have correctly configured external access, you should be able to get the same response using the external configuration URL (https://den.contoso.net/.well-known/configuration).

Stop Wayk Den, and wait for all microservices to stop:

[source, sh]
----
PS > Stop-WaykDen
----

=== User Management

User using Wayk Now client can log in to be authenticated with Wayk Den Server. The server, by default, will provide a Wayk Den ID to any user who wants to connect to it.

To authenticate user, Wayk Den can be configured to use a specific user group through LDAP integration. Two options are supported: Active Directory and JumpCloud. 

In order to fetch user and group information, a user with read-only LDAP access must first be created.

==== Active Directory

To integrate Active Directory, here is a list of the required parameters:

* LdapServerUrl: ldap://_ldap-server_
* LdapUsername, LdapPassword
* LdapUserGroup (optional)

It is important to specify the server IP since there is not DNS resolution in the docker container. The user used should be a user with only read-only access. A section below explains how to create a such user. Finally, the user group is not mandatory. If it is not specified, all users will be accepted. If it is specified, only users from that group will be able to be authenticated.

The following command will set LDAP property value for active directory. 

[source, sh]
----
Set-WaykDenConfig -LdapServerType ActiveDirectory -LdapUsername ldap-user@contoso.local -LdapPassword ldap-password -LdapServerUrl ldap://ldap-server -LdapUserGroup 'Domain Users'
---- 

===== User creation with read-only access

By default, a new user created in active directory has read-only access on the LDAP server. But that user is also member of the group Domain Users by default. Being member of that group is enough to be able to use that user and log on any domain's computer. To avoid that, we suggest to use a user who is not a member of Domain Users group and has only read-only access on the LDAP server. To do that, a few steps is needed.

First, a new group has to be created, let's say "Read-only Users". Then a new user can be created and added only to that group. After that, the new group can be set as primary group for the user. And finally, the user can be removed from the Domain Users group. This user should be used to configured WaykDen Server.

==== JumpCloud

https://jumpcloud.com/[JumpCloud] is a cloud service who help you to centralize user management. You can create users and groups then use the service call "LDAP-as-a-Service" to access those users and groups from WaykDen. You can read more on https://support.jumpcloud.com/customer/en/portal/articles/2439911-using-jumpcloud-s-ldap-as-a-service[how to use JumpCloud's LDAP-as-a-Service]. 

To integrate Jump Cloud with Wayk Den, here is a list of the required parameters:

* LdapServerUrl (ldaps://ldap.jumpcloud.com:636)
* LdapUsername, LdapPassword
* LdapBaseDn: Distinguished Name to retrieve users and groups
* LdapUserGroup (optional)

The LDAP server url should be set to ldaps://ldap.jumpcloud.com:636. JumpCloud provide a non secure access as well, but we don't recommand it. A user who can read the ldap directory should be created following steps https://support.jumpcloud.com/customer/en/portal/articles/2439911-using-jumpcloud-s-ldap-as-a-service#createuser[here]. The username has to be provided with the Distinguished Name (DN), something like `uid=_LDAP_BINDING_USER_,ou=Users,o=_YOUR_ORG_ID_,dc=jumpcloud,dc=com`. The base DN is similar and should be set to `ou=Users,o=_YOUR_ORG_ID_,dc=jumpcloud,dc=com`. Finally, a user group name can be specified to limit user to that group.

The following command will set LDAP property value for JumpCloud.

[source, sh]
----
Set-WaykDenConfig -LdapServerType JumpCloud -LdapUsername "uid=ldap-user,ou=Users,o=YOUR_ORG_ID,dc=jumpcloud,dc=com" -LdapPassword ldap-password -LdapServerUrl ldaps://ldap.jumpcloud.com:636 -LdapBaseDn "ou=Users,o=YOUR_ORG_ID,dc=jumpcloud,dc=com -LdapUserGroup wayk-users"
----


CHANGELOG.adoc
= Wayk Den Changelog

This document provides a list of notable changes introduced in Wayk Den by release.

== 2020.1.7 (2020-04-09)

  * Add docker isolation (hyperv, process) parameter for Windows containers
  * Fix problem when loading Wayk Den Web UI with null well-known endpoint
  * Fix system service cmdlet wrapper to use PowerShell encoded commands

== 2020.1.6 (2020-04-08)

  * Add Wayk Den and Jet relay system service registration (Windows only)
  * Add check to warn about DNS server being set to 127.0.0.1 on the host
  * Add check to warn about Symantec Endpoint Protection issues with docker
  * Add better error handling when Web UI cannot be loaded completely
  * Add "Update-WaykDen" command to force pulling latest container images
  * Fix issue with machine names containing special UTF-8 characters
  * Fix https listener on Linux due to broken paths in traefik.toml file

== 2020.1.5 (2020-03-25)

  * Fix Active Directory integration (LDAPS + simple bind)
  * Fix Devolutions Jet relay possible ghost sessions

== 2020.1.4 (2020-03-18)

  * Add support for web-based Wayk Now white-label bundle editor
  * Add support for Active Directory LDAPS integration with custom CA
  * Add "unlimited mode" for COVID-19 relief until September 19th, 2020

== 2020.1.3 (2020-02-19)

  * Add option to disable usage of a docker network
  * Add Devolutions Jet relay management commands
  * Add getting started guide with relay servers
  * Add getting started guide with ACME/letsencrypt

== 2020.1.2 (2020-02-06)

  * Fix support for Windows containers on Windows Server 2019
  * Add workaround for MongoDB Windows container lock file issue
  * Add Backup-WaykDenData/Restore-WaykDenData helper commands
  * Add getting started guide with an Azure virtual machine

== 2020.1.1 (2020-01-30)

  * Rewrite cmdlet in PowerShell instead of C#
  * Use YAML configuration files instead of LiteDB

== 2020.1.0 (2020-01-20)

  * Initial public release
  * Add initial Wayk Den web user interface
  * Add getting started guide with an Argo tunnel


WaykDen/Public/WaykDenConfig.ps1

. "$PSScriptRoot/../Private/CaseHelper.ps1"
. "$PSScriptRoot/../Private/YamlHelper.ps1"
. "$PSScriptRoot/../Private/TraefikHelper.ps1"
. "$PSScriptRoot/../Private/RandomGenerator.ps1"
. "$PSScriptRoot/../Private/CertificateHelper.ps1"

class WaykDenConfig
{
    # DenServer
    [string] $Realm
    [string] $ExternalUrl
    [string] $ListenerUrl
    [string] $ServerMode
    [int] $ServerCount
    [string] $DenServerUrl
    [string] $DenRouterUrl
    [string] $DenApiKey
    [bool] $ServerExternal

    # MongoDB
    [string] $MongoUrl
    [string] $MongoVolume
    [bool] $MongoExternal

    # Jet
    [string] $JetRelayUrl
    [string] $JetServerUrl

    # LDAP
    [string] $LdapServerUrl
    [string] $LdapServerIp
    [string] $LdapUsername
    [string] $LdapPassword
    [string] $LdapUserGroup
    [string] $LdapServerType
    [string] $LdapBaseDn
    [string] $LdapBindType
    [bool] $LdapCertificateValidation

    # Picky
    [string] $PickyUrl
    [string] $PickyApiKey
    [bool] $PickyExternal

    # Lucid
    [string] $LucidUrl
    [string] $LucidApiKey
    [string] $LucidAdminUsername
    [string] $LucidAdminSecret
    [bool] $LucidExternal

    # NATS
    [string] $NatsUrl
    [string] $NatsUsername
    [string] $NatsPassword
    [bool] $NatsExternal
    
    # Redis
    [string] $RedisUrl
    [string] $RedisPassword
    [bool] $RedisExternal

    # Docker
    [string] $DockerNetwork
    [string] $DockerPlatform
    [string] $DockerIsolation
    [string] $DockerHost
    [string] $SyslogServer
}

function Find-WaykDenConfig
{
    param(
        [string] $ConfigPath
    )

    if (-Not $ConfigPath) {
        $ConfigPath = Get-Location
    }

    if ($Env:DEN_CONFIG_PATH) {
        $ConfigPath = $Env:DEN_CONFIG_PATH
    }

    return $ConfigPath
}

function Set-WaykDenConfigPath
{
    param(
        [Parameter(Mandatory=$true,Position=0)]
        [string] $ConfigPath
    )

    $Env:DEN_CONFIG_PATH = $ConfigPath
}

function Get-WaykDenPath()
{
	[CmdletBinding()]
	param(
		[Parameter(Mandatory,Position=0)]
        [ValidateSet("ConfigPath","GlobalPath","LocalPath")]
		[string] $PathType
	)

    $DisplayName = "Wayk Den"
    $LowerName = "wayk-den"
    $CompanyName = "Devolutions"
	$HomePath = Resolve-Path '~'

	if (Get-IsWindows)	{
		$LocalPath = $Env:AppData + "\${CompanyName}\${DisplayName}";
		$GlobalPath = $Env:ProgramData + "\${CompanyName}\${DisplayName}"
	} elseif ($IsMacOS) {
		$LocalPath = "$HomePath/Library/Application Support/${DisplayName}"
		$GlobalPath = "/Library/Application Support/${DisplayName}"
	} elseif ($IsLinux) {
		$LocalPath = "$HomePath/.config/${LowerName}"
		$GlobalPath = "/etc/${LowerName}"
	}

	switch ($PathType) {
		'LocalPath' { $LocalPath }
		'GlobalPath' { $GlobalPath }
        'ConfigPath' { $GlobalPath }
		default { throw("Invalid path type: $PathType") }
	}
}

function Expand-WaykDenConfigKeys
{
    param(
        [WaykDenConfig] $Config
    )

    if (-Not $config.DenApiKey) {
        $config.DenApiKey = New-RandomString -Length 32
    }

    if (-Not $config.PickyApiKey) {
        $config.PickyApiKey = New-RandomString -Length 32
    }

    if (-Not $config.LucidApiKey) {
        $config.LucidApiKey = New-RandomString -Length 32
    }

    if (-Not $config.LucidAdminUsername) {
        $config.LucidAdminUsername = New-RandomString -Length 16
    }

    if (-Not $config.LucidAdminSecret) {
        $config.LucidAdminSecret = New-RandomString -Length 10
    }
}

function Expand-WaykDenConfig
{
    param(
        [WaykDenConfig] $Config
    )

    $DockerNetworkDefault = "den-network"
    $MongoUrlDefault = "mongodb://den-mongo:27017"
    $MongoVolumeDefault = "den-mongodata"
    $ServerModeDefault = "Private"
    $ListenerUrlDefault = "http://0.0.0.0:4000"
    $JetServerUrlDefault = "api.jet-relay.net:8080"
    $JetRelayUrlDefault = "https://api.jet-relay.net"
    $PickyUrlDefault = "http://den-picky:12345"
    $LucidUrlDefault = "http://den-lucid:4242"
    $DenServerUrlDefault = "http://den-server:10255"
    $DenRouterUrlDefault = "http://den-server:4491"

    if (-Not $config.DockerNetwork) {
        $config.DockerNetwork = $DockerNetworkDefault
    }

    if (($config.DockerNetwork -Match "none") -and $config.DockerHost) {
        $MongoUrlDefault = $MongoUrlDefault -Replace "den-mongo", $config.DockerHost
        $PickyUrlDefault = $PickyUrlDefault -Replace "den-picky", $config.DockerHost
        $LucidUrlDefault = $LucidUrlDefault -Replace "den-lucid", $config.DockerHost
        $DenServerUrlDefault = $DenServerUrlDefault -Replace "den-server", $config.DockerHost
        $DenRouterUrlDefault = $DenRouterUrlDefault -Replace "den-server", $config.DockerHost
    }

    if (-Not $config.DockerPlatform) {
        if (Get-IsWindows) {
            $config.DockerPlatform = "windows"
        } else {
            $config.DockerPlatform = "linux"
        }
    }

    if (-Not $config.ServerMode) {
        $config.ServerMode = $ServerModeDefault
    }

    if (-Not $config.ServerCount) {
        $config.ServerCount = 1
    }

    if (-Not $config.ListenerUrl) {
        $config.ListenerUrl = $ListenerUrlDefault
    }

    if (-Not $config.MongoUrl) {
        $config.MongoUrl = $MongoUrlDefault
    }

    if (-Not $config.MongoVolume) {
        $config.MongoVolume = $MongoVolumeDefault
    }

    if (-Not $config.JetServerUrl) {
        $config.JetServerUrl = $JetServerUrlDefault
    }

    if (-Not $config.JetRelayUrl) {
        $config.JetRelayUrl = $JetRelayUrlDefault
    }

    if (-Not $config.PickyUrl) {
        $config.PickyUrl = $PickyUrlDefault
    }

    if (-Not $config.LucidUrl) {
        $config.LucidUrl = $LucidUrlDefault
    }

    if (-Not $config.DenServerUrl) {
        $config.DenServerUrl = $DenServerUrlDefault
    }

    if (-Not $config.DenRouterUrl) {
        $config.DenRouterUrl = $DenRouterUrlDefault
    }
}

function Export-TraefikToml()
{
    param(
        [string] $ConfigPath
    )

    $ConfigPath = Find-WaykDenConfig -ConfigPath:$ConfigPath

    $config = Get-WaykDenConfig -ConfigPath:$ConfigPath
    Expand-WaykDenConfig $config

    $TraefikPath = Join-Path $ConfigPath "traefik"
    New-Item -Path $TraefikPath -ItemType "Directory" -Force | Out-Null

    $TraefikTomlFile = Join-Path $TraefikPath "traefik.toml"

    $TraefikToml = New-TraefikToml -Platform $config.DockerPlatform `
        -ListenerUrl $config.ListenerUrl `
        -LucidUrl $config.LucidUrl `
        -PickyUrl $config.PickyUrl `
        -DenRouterUrl $config.DenRouterUrl `
        -DenServerUrl $config.DenServerUrl

    Set-Content -Path $TraefikTomlFile -Value $TraefikToml
}

function Export-HostInfo()
{
    param(
        [string] $ConfigPath,
        [PSCustomObject] $HostInfo
    )

    $ConfigPath = Find-WaykDenConfig -ConfigPath:$ConfigPath

    $config = Get-WaykDenConfig -ConfigPath:$ConfigPath
    Expand-WaykDenConfig $config

    $DenServerPath = Join-Path $ConfigPath "den-server"
    New-Item -Path $DenServerPath -ItemType "Directory" -Force | Out-Null

    $JsonValue = $($HostInfo | ConvertTo-Json)
    $HostInfoFile = Join-Path $DenServerPath "host_info.json"
    Set-Content -Path $HostInfoFile -Value $JsonValue -Force
}

function New-WaykDenConfig
{
    [CmdletBinding()]
    param(
        [string] $ConfigPath,
    
        # Server
        [Parameter(Mandatory=$true)]
        [string] $Realm,
        [Parameter(Mandatory=$true)]
        [string] $ExternalUrl,
        [string] $ListenerUrl,
        [string] $ServerMode,
        [int] $ServerCount,
        [string] $DenServerUrl,
        [string] $DenRouterUrl,
        [string] $DenApiKey,
        [bool] $ServerExternal,

        # MongoDB
        [string] $MongoUrl,
        [string] $MongoVolume,
        [bool] $MongoExternal,

        # Jet
        [string] $JetRelayUrl,

        # LDAP
        [string] $LdapServerUrl,
        [string] $LdapServerIp,
        [string] $LdapUsername,
        [string] $LdapPassword,
        [string] $LdapUserGroup,
        [string] $LdapServerType,
        [string] $LdapBaseDn,
        [string] $LdapBindType,
        [bool] $LdapCertificateValidation,

        # Picky
        [string] $PickyUrl,
        [string] $PickyApiKey,
        [bool] $PickyExternal,

        # Lucid
        [string] $LucidUrl,
        [string] $LucidApiKey,
        [string] $LucidAdminUsername,
        [string] $LucidAdminSecret,
        [bool] $LucidExternal,

        # NATS
        [string] $NatsUrl,
        [string] $NatsUsername,
        [string] $NatsPassword,
        [bool] $NatsExternal,
        
        # Redis
        [string] $RedisUrl,
        [string] $RedisPassword,
        [bool] $RedisExternal,

        # Docker
        [string] $DockerNetwork,
        [string] $DockerPlatform,
        [string] $DockerIsolation,
        [string] $DockerHost,
        [string] $SyslogServer,

        [switch] $Force
    )

    $ConfigPath = Find-WaykDenConfig -ConfigPath:$ConfigPath

    New-Item -Path $ConfigPath -ItemType "Directory" -Force | Out-Null
    $ConfigFile = Join-Path $ConfigPath "wayk-den.yml"

    $DenServerPath = Join-Path $ConfigPath "den-server"
    $DenPublicKeyFile = Join-Path $DenServerPath "den-public.pem"
    $DenPrivateKeyFile = Join-Path $DenServerPath "den-private.key"
    New-Item -Path $DenServerPath -ItemType "Directory" -Force | Out-Null

    if (!((Test-Path -Path $DenPublicKeyFile -PathType "Leaf") -and
          (Test-Path -Path $DenPrivateKeyFile -PathType "Leaf"))) {
            $KeyPair = New-RsaKeyPair -KeySize 2048
            Set-Content -Path $DenPublicKeyFile -Value $KeyPair.PublicKey -Force
            Set-Content -Path $DenPrivateKeyFile -Value $KeyPair.PrivateKey -Force
    }

    $config = [WaykDenConfig]::new()
    
    $properties = [WaykDenConfig].GetProperties() | ForEach-Object { $_.Name }
    foreach ($param in $PSBoundParameters.GetEnumerator()) {
        if ($properties -Contains $param.Key) {
            $config.($param.Key) = $param.Value
        }
    }

    Expand-WaykDenConfigKeys -Config:$config

    ConvertTo-Yaml -Data (ConvertTo-SnakeCaseObject -Object $config) -OutFile $ConfigFile -Force:$Force

    Export-TraefikToml -ConfigPath:$ConfigPath
}

function Set-WaykDenConfig
{
    [CmdletBinding()]
    param(
        [string] $ConfigPath,
    
        # Server
        [string] $Realm,
        [string] $ExternalUrl,
        [string] $ListenerUrl,
        [string] $ServerMode,
        [int] $ServerCount,
        [string] $DenServerUrl,
        [string] $DenRouterUrl,
        [string] $DenApiKey,
        [bool] $ServerExternal,

        # MongoDB
        [string] $MongoUrl,
        [string] $MongoVolume,
        [bool] $MongoExternal,

        # Jet
        [string] $JetRelayUrl,

        # LDAP
        [string] $LdapServerUrl,
        [string] $LdapServerIp,
        [string] $LdapUsername,
        [string] $LdapPassword,
        [string] $LdapUserGroup,
        [string] $LdapServerType,
        [string] $LdapBaseDn,
        [string] $LdapBindType,
        [bool] $LdapCertificateValidation,

        # Picky
        [string] $PickyUrl,
        [string] $PickyApiKey,
        [bool] $PickyExternal,

        # Lucid
        [string] $LucidUrl,
        [string] $LucidApiKey,
        [string] $LucidAdminUsername,
        [string] $LucidAdminSecret,
        [bool] $LucidExternal,

        # NATS
        [string] $NatsUrl,
        [string] $NatsUsername,
        [string] $NatsPassword,
        [bool] $NatsExternal,
        
        # Redis
        [string] $RedisUrl,
        [string] $RedisPassword,
        [bool] $RedisExternal,

        # Docker
        [string] $DockerNetwork,
        [string] $DockerPlatform,
        [string] $DockerIsolation,
        [string] $DockerHost,
        [string] $SyslogServer,

        [switch] $Force
    )

    $ConfigPath = Find-WaykDenConfig -ConfigPath:$ConfigPath

    $config = Get-WaykDenConfig -ConfigPath:$ConfigPath

    New-Item -Path $ConfigPath -ItemType "Directory" -Force | Out-Null
    $ConfigFile = Join-Path $ConfigPath "wayk-den.yml"

    $properties = [WaykDenConfig].GetProperties() | ForEach-Object { $_.Name }
    foreach ($param in $PSBoundParameters.GetEnumerator()) {
        if ($properties -Contains $param.Key) {
            $config.($param.Key) = $param.Value
        }
    }

    Expand-WaykDenConfigKeys -Config:$config
 
    # always force overwriting wayk-den.yml when updating the config file
    ConvertTo-Yaml -Data (ConvertTo-SnakeCaseObject -Object $config) -OutFile $ConfigFile -Force

    Export-TraefikToml -ConfigPath:$ConfigPath
}

function Get-WaykDenConfig
{
    [CmdletBinding()]
    [OutputType('WaykDenConfig')]
    param(
        [string] $ConfigPath,
        [switch] $Expand
    )

    $ConfigPath = Find-WaykDenConfig -ConfigPath:$ConfigPath

    $ConfigFile = Join-Path $ConfigPath "wayk-den.yml"
    $ConfigData = Get-Content -Path $ConfigFile -Raw -ErrorAction Stop
    $yaml = ConvertFrom-Yaml -Yaml $ConfigData -UseMergingParser -AllDocuments -Ordered

    $config = [WaykDenConfig]::new()

    [WaykDenConfig].GetProperties() | ForEach-Object {
        $Name = $_.Name
        $snake_name = ConvertTo-SnakeCase -Value $Name
        if ($yaml.Contains($snake_name)) {
            if ($yaml.$snake_name -is [string]) {
                if (![string]::IsNullOrEmpty($yaml.$snake_name)) {
                    $config.$Name = ($yaml.$snake_name).Trim()
                }
            } else {
                $config.$Name = $yaml.$snake_name
            }
        }
    }

    if ($Expand) {
        Expand-WaykDenConfig $config
    }

    return $config
}

Export-ModuleMember -Function New-WaykDenConfig, Set-WaykDenConfig, Get-WaykDenConfig, `
    Set-WaykDenConfigPath, Get-WaykDenPath
